src/v/cloud_storage/remote_partition.cc

```diff
@@ -743,13 +749,31 @@ void remote_partition::maybe_sync_with_manifest() {
 
             std::visit([this](auto&& p) { p->offload(this); }, prev_it->second);
         }
-        auto res = _segments.insert_or_assign(
-          o, offloaded_segment_state(meta.second.base_offset));
-        if (res.second) {
-            _probe.segment_added();
+        auto spath = _manifest.generate_segment_path(meta.second);
+        if (auto it = materialized_segments.find(spath);
+            it != materialized_segments.end()) {
+            auto ms = std::move(it->second);
+            _segments.insert_or_assign(o, std::move(ms));
+            materialized_segments.erase(it);
+        } else {
+            auto res = _segments.insert_or_assign(
+              o, offloaded_segment_state(meta.second.base_offset));
+            // With offloaded segments we can't distinguish between compacted
+            // and non compacted segments with the same base offset. Because of
+            // that all offloaded segments has to be recreated every time. It
+            // makes this metric useless but future update will make it
+            // completely obsolete since we will no longer have segments in
+            // offloaded state.
+            if (res.second) {
+                _probe.segment_added();
```

Comment:
nit/followup: perhaps we should go ahead and remove the metric now, rather than leaving it in a strange state?

Reply:
yep